src/goto-programs/goto_convert_functions.cpp

```diff
@@ -354,9 +606,108 @@ void goto_convert_functionst::wallop_type(
   // And finally perform renaming.
   symbolt *s = context.find_symbol(name);
   rename_types(s->type, *s, sname);
-  deps.clear();
 }
 
+using node_id = context_type_grapht::node_id;
+using edge = context_type_grapht::edge;
+
+/* Support node_id hashing using types in std */
+namespace std
+{
+template <>
+struct hash<node_id> : hash<size_t>
+{
+  static_assert(sizeof(node_id) == sizeof(size_t));
+
+  size_t operator()(const node_id &v) const noexcept
+  {
+    size_t w;
+    memcpy(&w, &v, sizeof(w));
+    return hash<size_t>::operator()(w); // this is basically a no-op
+  }
+};
+} // namespace std
+
+namespace
+{
+struct sccst /* strongly connected components */
+{
+  const context_type_grapht &G;
+
+  struct tarjan_data
+  {
+    size_t index = 0;
+    size_t lowlink;
+    bool on_stack;
+  };
+
+  node_map<tarjan_data> data;
+  std::vector<node_id> stack;
+  size_t index = 0;
+
+  std::unordered_set<node_id> large_scc_syms;
+
+  sccst(const context_type_grapht &G) : G(G), data(G)
+  {
+  }
+
+  void tarjan(node_id v) /* Tarjan's 1972 algorithm to compute SCCs */
```

Comment:
Can you add the paper's title and publication venue here?

Reply:
done.